### PR TITLE
CHANGE (GraphEngine): @W-12439992@: Applies limit on how far Path Expansion registry data can grow to prevent OutOfMemory error

### PR DESCRIPTION
**Pull request by rmohan20**
1. New `SFGE_PATH_EXPANSION_LIMIT` env variable to enforce an upper limit value. -1 indicates that no limit will be imposed. By default, the limit is calculated as the number of average-sized ApexPathExpander instances that can fit into 50% of the max heap space allocated.
2. `RegistryDataLimitCalculator` handles the default value calculation mentioned above.
3. Any time the limit is reached, a PathExpansionLimitReached error/violation is thrown to user for the specific path entry. This allows the process to go on for other entry points.
4. New tests for the limit changes.

What's not covered?
1. UI changes from NodeJS layer.

### CODE REVIEW

File: sfge/src/main/java/com/salesforce/config/UserFacingMessages.java [hunk 1]
```diff
@@ -20,6 +20,8 @@ public static final class RuleViolationTemplates {
         // Format: First %s is either "abstract class" or "interface".
         //         Second %s is the name of a class or interface.
         public static final String UNIMPLEMENTED_TYPE_RULE = "Extend, implement, or delete %s %s";
+        public static final String LIMIT_REACHED_VIOLATION_MESSAGE =
+                "%s. Graph Engine is preemptively failing this path to prevent the occurrence of OutOfMemory error. Rerun Graph Engine targeting this entry method with a larger heap space.";
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/config/UserFacingMessages.java:24]: @teresa-allen-sfdc Could you review this message please?

File: sfge/src/main/java/com/salesforce/config/UserFacingMessages.java [hunk 2]
```diff
@@ -39,6 +41,9 @@ public static final class RuleViolationTemplates {
     public static final String VARIABLE_DECLARED_MULTIPLE_TIMES =
             "Rename or remove reused variable to proceed with analysis: %s,%s:%d";
 
+    public static final String INSUFFICIENT_HEAP_SPACE =
+            "Insufficient heap space (%d bytes) to execute Graph Engine. Increase heap space using --sfgejvmargs option and retry.";
+
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/config/UserFacingMessages.java:46]: @teresa-allen-sfdc Could you review this message please?

File: sfge/src/main/java/com/salesforce/config/UserFacingMessages.java [hunk 3]
```diff
@@ -54,4 +59,7 @@ public static final class RuleViolationTemplates {
     public static final String FIX_COMPILATION_ERRORS = "Fix compilation errors in %s and retry";
 
     public static final String EXCEPTION_FORMAT_TEMPLATE = "%s, Caused by:\n%s";
+
+    public static final String PATH_EXPANSION_LIMIT_REACHED =
+            "Reached path expansion upper limit (%d)";
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/config/UserFacingMessages.java:64]: @teresa-allen-sfdc This is another user-facing message.

File: sfge/src/main/java/com/salesforce/graph/ops/expander/PathExpansionRegistry.java [hunk 4]
```diff
@@ -56,9 +61,34 @@ private static class ForkEventRegistryData extends RegistryData<ForkEvent> {
...
+        final int pathExpansionLimit;
+        final boolean isPathExpansionLimitSet;
+
+        ApexPathExpanderRegistryData() {
+            pathExpansionLimit = SfgeConfigProvider.get().getPathExpansionLimit();
+            isPathExpansionLimitSet = pathExpansionLimit > 0;
+        }
+
+        @Override
+        public void validateAndPut(ApexPathExpander instance) {
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/ops/expander/PathExpansionRegistry.java:79]: We check the registry size each time before adding more items. If there's no limit set, we directly add the item.

File: sfge/src/main/java/com/salesforce/graph/ops/registry/PathExpansionLimitReachedException.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package com.salesforce.graph.ops.registry;
+
+import com.salesforce.config.UserFacingMessages;
+import com.salesforce.exception.SfgeRuntimeException;
+
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/ops/registry/PathExpansionLimitReachedException.java:5]: Will add javadoc

File: sfge/src/main/java/com/salesforce/graph/ops/registry/RegistryDataLimitCalculator.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
+package com.salesforce.graph.ops.registry;
+
+import com.salesforce.config.UserFacingMessages;
+import com.salesforce.exception.UserActionException;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/** Helps calculate the limit to enforce on the registry for a given type of {@link Registrable}. */
+public class RegistryDataLimitCalculator {
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/ops/registry/RegistryDataLimitCalculator.java:9]: Handles limit calculation.

File: sfge/src/main/java/com/salesforce/rules/Violation.java [hunk 7]
```diff
@@ -302,6 +303,33 @@ public int compareTo(Violation otherViolation) {
...
+
+            // TODO: reconsider the name and category. We want users to take followup action.
+            this.ruleName = "LimitReached";
+            this.category = INTERNAL_ERROR_CATEGORY;
+            this.description = "";
+            this.severity = AbstractRule.SEVERITY.LOW.code;
+            // NOTE: For now, we're hardcoding the URL to the URL for the Apex FLS violation rule,
+            // since that's the only rule that can throw this violation. This may change in a future
+            // release.
+            this.url = ApexFlsViolationRule.URL;
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/rules/Violation.java:329]: @teresa-allen-sfdc Once we have a separate section in our docs on how to handle this situation, we can update the link here.

**Review by jfeingold35 — APPROVED** [review]

**Pull request merged**